### PR TITLE
Don't remove metadata when other instance of service (with same name) remains

### PR DESCRIPTION
**Pull request by amochtar**

### CODE REVIEW

File: consulmeta/consulmeta.go [hunk 1]
```diff
@@ -109,12 +109,19 @@ func (r *ConsulMetaAdapter) buildCheck(service *bridge.Service) *consulapi.Agent
 
 func (r *ConsulMetaAdapter) Deregister(service *bridge.Service) error {
 	var err error
-	path := r.path[1:] + "/" + service.Name
-	_, err = r.client.KV().DeleteTree(path, nil)
+
+	err = r.client.Agent().ServiceDeregister(service.ID)
 	if err != nil {
 		return err
```
Comment (simonvanderveldt) [inline consulmeta/consulmeta.go]: Maybe add a log message here as well?

File: consulmeta/consulmeta.go [hunk 2]
```diff
@@ -64,19 +64,19 @@ func (r *ConsulMetaAdapter) Register(service *bridge.Service) error {
...
 		_, err = r.client.KV().Put(&consulapi.KVPair{Key: path, Value: []byte(v)}, nil)
 	}
 	if err != nil {
+		log.Println("consulmeta: failed to register metadata for service:", err)
+	}
+
+	err = r.client.Agent().ServiceRegister(registration)
+	if err != nil {
 		log.Println("consulmeta: failed to register service:", err)
+		return err
```
Comment (simonvanderveldt) [inline consulmeta/consulmeta.go]: Why are you returning `err` here and not on line 73/74?

File: consulmeta/consulmeta.go [hunk 3]
```diff
@@ -109,12 +109,19 @@ func (r *ConsulMetaAdapter) buildCheck(service *bridge.Service) *consulapi.Agent
...
-	_, err = r.client.KV().DeleteTree(path, nil)
+
+	err = r.client.Agent().ServiceDeregister(service.ID)
 	if err != nil {
 		return err
 	}
-	return r.client.Agent().ServiceDeregister(service.ID)
+
+	// Only remove KV data when there are no more services with the same name
+	var services, _, _ = r.client.Catalog().Service(service.Name, "", nil)
```
Comment (simonvanderveldt) [inline consulmeta/consulmeta.go]: :o is there no way to directly check if a service exists? something like `service.exists('servicename')`?

File: consulmeta/consulmeta.go (hunk 2 above)
Comment (amochtar) [inline consulmeta/consulmeta.go]: Good question. I meant to change that, but apparently forgot :)

File: consulmeta/consulmeta.go (hunk 3 above)
Comment (amochtar) [inline consulmeta/consulmeta.go]: Nope, this appears to be the only way...

Comment (simonvanderveldt) [inline consulmeta/consulmeta.go]: :(

**Comment by simonvanderveldt** [comment]
LGTM code wise. How can we test/verify if this works/doesn't break?

**Comment by amochtar** [comment]
@simonvanderveldt I've created a second automated build for this branch, so you can run `amochtar/registrator:multi-instance` and verify

**Comment by simonvanderveldt** [comment]
Ran both the current `master` and this branch locally.
When killing 1 of 2 services the following happens:
- `master`: the KV tree for the service is removed
- this branch: The KV tree for the service stays

When killing both services the KV tree for this service is removed

**Pull request merged**